File: _Demo_and_Tests/Objective-C/AlgorithmsTests/LinearSearchEqualObjectsTests.m
//
//  LinearSearchEqualObjectsTests.m
//  Algorithms
//

#import <XCTest/XCTest.h>
#import "LinearSearch.h"

@interface LinearSearchEqualObjectsTests : XCTestCase

@property (strong, nonatomic) NSArray *array;

@end

@implementation LinearSearchEqualObjectsTests

- (void)setUp {
    [super setUp];
}

- (void)tearDown {
    [super tearDown];
}

- (NSArray *)array {
    if (!_array) {
        _array = @[@(-1), @0, @1, @1,
                   @(-1.5), @0.5, @1.5, @1.5];
    }

    return _array;
}

#pragma mark - Integers (point to the same memory address)

- (void)testInteger {
    NSUInteger resultIndex = [LinearSearch indexOfEqualObject:@1.0 inArray:self.array];

    XCTAssert(resultIndex == 2);
}

- (void)testIntegers {
    NSArray *resultArray = [LinearSearch indexesOfEqualObjects:@1.0 inArray:self.array];
    NSArray *expectedArray = @[@2, @3];

    XCTAssertEqualObjects(resultArray, expectedArray);
}

#pragma mark - Floats (point to the same memory address)

- (void)testFloat {
    NSUInteger resultIndex = [LinearSearch indexOfEqualObject:@1.5 inArray:self.array];

    XCTAssert(resultIndex == 6);
}

- (void)testFloats {
    NSArray *resultArray = [LinearSearch indexesOfEqualObjects:@1.5 inArray:self.array];
    NSArray *expectedArray = @[@6, @7];

    XCTAssertEqualObjects(resultArray, expectedArray);
}

#pragma mark - Not Included

- (void)testNotIncludedObject {
    NSUInteger resultIndex = [LinearSearch indexOfEqualObject:@100 inArray:self.array];

    XCTAssert(resultIndex == NSNotFound);
}

- (void)testNotIncludedObjects {
    NSArray *resultArray = [LinearSearch indexesOfEqualObjects:@100 inArray:self.array];

    XCTAssertEqualObjects(resultArray, @[]);
}

#pragma mark - Nil Search Object

- (void)testNilSearchObject {
    NSUInteger resultIndex = [LinearSearch indexOfEqualObject:nil inArray:self.array];

    XCTAssert(resultIndex == NSNotFound);
}

- (void)testNilSearchObjects {
    NSArray *resultArray = [LinearSearch indexesOfEqualObjects:nil inArray:self.array];

    XCTAssertEqualObjects(resultArray, @[]);
}

#pragma mark - Nil Array

- (void)testNilArrayForObject {
    NSUInteger resultIndex = [LinearSearch indexOfEqualObject:@YES inArray:nil];

    XCTAssert(resultIndex == NSNotFound);
}

- (void)testNilArrayForObjects {
    NSArray *resultArray = [LinearSearch indexesOfEqualObjects:@YES inArray:nil];
    
    XCTAssertEqualObjects(resultArray, @[]);
}

@end

File: _Demo_and_Tests/Swift/AlgorithmsTests/LinearSearchEqualObjectsTests.swift
//
//  LinearSearchEqualObjectsTests.swift
//  Algorithms
//

import XCTest

class LinearSearchEqualObjectsTests: XCTestCase {

    lazy var objectsArray: [AnyObject] = [NSNumber.init(value: -1), NSNumber.init(value: 0), NSNumber.init(value: 1), NSNumber.init(value: 1),
                                          NSNumber.init(value: -1.5), NSNumber.init(value: 0.5), NSNumber.init(value: 1.5), NSNumber.init(value: 1.5)]

    override func setUp() {
        super.setUp()
    }

    override func tearDown() {
        super.tearDown()
    }

    // MARK: - Integers (point to the same memory adress)

    func testInteger() {
        let resultIndex = LinearSearch.index(ofEqualObject: NSNumber.init(value: 1), in: objectsArray)

        XCTAssert(resultIndex == 2)
    }

    func testIntegers() {
        let resultArray = LinearSearch.indexes(ofEqualObject: NSNumber.init(value: 1), in: objectsArray)

        XCTAssert(resultArray == [2, 3])
    }

    // MARK: - Floats

    func testFloat() {
        let resultIndex = LinearSearch.index(ofEqualObject: NSNumber.init(value: 1.5), in: objectsArray)

        XCTAssert(resultIndex == 6)
    }

    func testFloats() {
        let resultArray = LinearSearch.indexes(ofEqualObject: NSNumber.init(value: 1.5), in: objectsArray)

        XCTAssert(resultArray == [6, 7])
    }

    // MARK: - Not Included

    func testNotIncludedObject() {
        let resultIndex = LinearSearch.index(ofEqualObject: NSNumber.init(value: 100), in: objectsArray)

        XCTAssert(resultIndex == nil)
    }

    func testNotIncludedObjects() {
        let resultArray = LinearSearch.indexes(ofEqualObject: NSNumber.init(value: 100), in: objectsArray)

        XCTAssert(resultArray == [])
    }

    // MARK: - Nil Search Object

    func testNilSearchObject() {
        let resultIndex = LinearSearch.index(ofEqualObject: nil, in: objectsArray)

        XCTAssert(resultIndex == nil)
    }

    func testNilSearchObjects() {
        let resultArray = LinearSearch.indexes(ofEqualObject: nil, in: objectsArray)

        XCTAssert(resultArray == [])
    }

    // MARK: - Nil Array

    func testNilArrayForObject() {
        let resultIndex = LinearSearch.index(ofEqualObject: NSNumber.init(value: true), in: nil)

        XCTAssert(resultIndex == nil)
    }

    func testNilArrayForObjects() {
        let resultArray = LinearSearch.indexes(ofEqualObject: NSNumber.init(value: true), in: nil)
        
        XCTAssert(resultArray == [])
    }

}
